### DISCUSSION TITLE
Proposed implementation of an FLTK text widget adapted to all scripts, RtL included

### DISCUSSION THREAD

**Discussion by ManoloFLTK** [fltk/fltk#1174] [Development]
Proposed implementation of an FLTK text input widget adapted to all scripts.

This discussion is related to "Middle Eastern languages (Arabic) internationalization" (#1063). 

Available for evaluation in branch `text_widget` of `https://github.com/ManoloFLTK/fltk`
Download and build demo app `examples/native-input` as follows:
```
git clone https://github.com/ManoloFLTK/fltk   fltk-manolo
cd fltk-manolo
git checkout text_widget
mkdir build
cd build
cmake -DFLTK_BUILD_EXAMPLES=1 ..
make -j4 native-input
```

**TL;DR:** FLTK is extended with class `Fl_Native_Input` which supports text input in any kind of script, right-to-left included. The public API of this class is the same  as that of class `Fl_Input_`. Derived class `Fl_Native_Multiline_Input` adds multiple-line support, line wrapping control and scrollbars.

Use a dedicated demo program - `examples/native-input` - to exercize the new widget:
- open with menu item "File/Open" a text file, possibly containing right-to-left text.
The app window contains a new multiple-line input widget, a new single-line widget with some Arabic text, and a usual Fl_Input.
- experiment with all three widgets: input and edit text, scroll text, text selection, copy/paste/cut, drag-n-drop, undo, redo.
- change text font or size or color, inactivate, re-activate with the Format menu.
- resize window and see line wrapping
- open another text file containing left-to-right text, and apply menu item "Edit/RtoL<->LtoR" if necessary to display it correctly.
- print all that to pdf with menu item "File/Print to PDF".

The idea is to arrange for FLTK to use an input widget each platform has in place natively that is able to handle all scripts, right-to-left included. This way, FLTK doesn't need to re-implement the tasks of text edition and line wrapping control that become very complex to support all scripts. FLTK just inherits such features from the platform software. It becomes necessary to allow FLTK to interact properly with a non-FLTK widget.

For macOS, such native widget is `NSTextView` (see https://developer.apple.com/documentation/appkit/nstextview).
For Windows, that is an `Edit control` (see https://learn.microsoft.com/en-us/windows/win32/controls/edit-controls).
For Linux/Unix, X11 itself doesn't offer such a widget, but GTK does with `Gtk.Textview` (see https://docs.gtk.org/gtk3/class.TextView.html). The exact same code works under X11 and under Wayland via `Fl_Cairo_Graphics_Driver` common to these 2 platforms.

A new FLTK widget named `Fl_Native_Input` is added. It's a child of `Fl_Input`.
The public API of `Fl_Native_Input` is exactly that of `Fl_Input_`. The new widget can therefore be easily used by applications as a replacement of `Fl_Input` or `Fl_Multiline_Input` widgets.

The new widget hierarchy in use is
```
Fl_Input
    Fl_Native_Input         <== same public API as Fl_Input_, for single-line text input
        Fl_Native_Multiline_Input  <== adds multiple-line support, line wrapping control,
                       vertical scrollbar, and horizontal scrollbar in absence of line wrapping.
 ```
The new code introduces a new dependency, to GTK3, only under Linux/Unix. 
A new CMake option `FLTK_OPTION_NATIVE_INPUT` that's ON by default is defined. If set OFF, classes `Fl_Native_Input` and  `Fl_Native_Multiline_Input` are not added to the FLTK library.

Source and header files added to the git repository:
-  `Fl_Native_Input.{cxx,H}` to support new classes `Fl_Native_Input` and `Fl_Native_Multiline_Input`.
-  `src/Fl_Native_Input_Driver.H` to declare internal class `Fl_Native_Input_Driver` that bridges public class `Fl_Native_Input` to its platform-specific implementation.
-  `src/drivers/XXX/Fl_XXX_Native_Input_Driver.cxx` contains the source code for the new widget specific to platform XXX. The extension is .mm under macOS.

In addition to new source files and new headers, a single change to the FLTK source code is necessary.  This change  doesn't break the FLTK 1.4.0 ABI. It's macOS-specific and occurs in file `src/Fl_cocoa.mm`. It exists because FLTK must be able to give focus to `NSTextView` objects which is a new requirement brought by class `Fl_Native_Input`.

Platform-specific implementation details

**macOS**: `Fl_Native_Input` is implemented adding an `NSScrollView` as a subview of the top window's contentView and adding an `NSTextView` as the `NSScrollView`'s document view. This gives support for all needed scrollbars of `Fl_Native_Multiline_Input`. It requires one change in method `performKeyEquivalent:` of file `src/Fl_cocoa.mm` to allow a native subview to receive shortcuts. Previously, only the contentView of a top-level window would be a window's first responder and receive events. FLTK's event mechanism would transfer the event to child windows and widgets. When the focus goes to an `Fl_Native_Input` object, the new code makes the `NSTextView` the first responder of its containing top window. Once all this is in place, text input is completely handled by the `NSTextView` native object which receives and processes events, and draws the text on screen. 
Some more code allows FLTK to control what happens when multiple-line text is pasted or dropped to a single-line widget. End-of-lines are replaced by ␍. Conversely, ␍ are replaced by end-of-lines when text is copied from a single-line widget. An `Fl_Native_Input` object is drawn to any non-display `Fl_Surface_Device` by capturing the widget from the display as a `CGImage` and drawing that `CGImage` to the `Fl_Surface_Device`.

**Windows**: `Fl_Native_Input` is implemented creating an Edit Control object which has type HWND and adding it as a child of its containing `Fl_Window`. Right after the creation of the Edit control, the code replaces its parent window's `WndProc` by function `fltk_wnd_proc_plus_focus()` and memorizes the initial `WndProc` of that window under the name `fltk_wnd_proc()`. This allows to transfer focus to the control when necessary as follows. The OS sends the `WM_COMMAND` message and the `EN_SETFOCUS` wParam to the control's parent window when the edit control receives the keyboard focus. This message is received by function `fltk_wnd_proc_plus_focus()` that calls `Fl_Widget::take_focus()` to give focus to the control and then forwards the message to `fltk_wnd_proc()`. Function `fltk_wnd_proc_plus_focus()` processes also `WM_CTLCOLOREDIT` messages to set the Edit Control text and background colors. Transmission of events to the Edit Control is done replacing its `WndProc` by another one named `fltk_edit_wnd_proc()` and memorizing the Control's initial `WndProc` under the name `win32_edit_wnd_proc()`.  Function `fltk_edit_wnd_proc()` forwards all messages it receives, except `WM_KEYDOWN` messages, to `win32_edit_wnd_proc()`. This ensures that text-editing operations are handled by the Edit Control which also draws the managed text adequately on screen. As for each `WM_KEYDOWN` message, `fltk_edit_wnd_proc()` checks whether it results from a Tab or Escape key press or from a Ctrl+key other than Ctrl-C or Ctrl-V. If so, that message is sent to FLTK's default `WndProc` function - `fltk_wnd_proc()`. This ensures all FLTK shortcuts including menu item shortcuts as well as Tab and Escape keypresses are catched and processed adequately. Otherwise, the `WM_KEYDOWN` message is forwarded to `win32_edit_wnd_proc()` and processed as expected from an Edit control. An `Fl_Native_Input` object is drawn to any non-display `Fl_Surface_Device` by capturing the widget from the display as an `Fl_RGB_Image` and drawing the `Fl_RGB_Image` to the `Fl_Surface_Device`.

**Wayland/X11**: Implementation of class `Fl_Native_Input` for this platform is more challenging than for macOS and Windows because the GTK object used - `Gtk.Textview` - is primarily conceived to be added to a GTK window and to interact with the GTK event loop. But, the FLTK library doesn't create GTK top-level windows and doesn't run the GTK event loop. FLTK doesn't change how it manages focus nor how it receives events to support class `Fl_Native_Input` under Wayland/X11. Member function `Fl_Native_Input::handle(int)` receives all FLTK events sent to the new widget and converts each into calls to GTK functions that put the `Gtk.Textview` in the adequate state. The GTK library contains the `Gtk.OffscreenWindow` class which purpose is documented as follows: 
>> The idea is to take a widget and manually set the state of it, add it to a Gtk.OffscreenWindow and then retrieve the snapshot as a cairo_surface_t or GdkPixbuf.

That's what FLTK does. FLTK implements an `Fl_Native_Input` object with a `Gtk.OffscreenWindow` containing a `Gtk.ScrolledWindow` which contains itself a `Gtk.Textview`. When the virtual member method `Fl_Native_Input::draw()` runs, FLTK instructs the GTK library to draw the `Gtk.ScrolledWindow` object with function `gtk_widget_draw()` to the current window's Cairo surface with the appropriate translation. The result is that the `Gtk.Textview` appears on the display with its correct text content. An `Fl_Native_Input` object is made of a central text area drawn as just seen, optional vertical and horizontal scrollers that are normal FLTK widgets, and the widget's frame. The vertical scroller is placed at left of the central area when the widget contains right-to-left text. This implementation fully supports undo/redo operations with a method that is a straight copy of the undo/redo mechanism of class `Fl_Input_` after adaptation to GTK objects where locations inside the text handled by the `Gtk.Textview` object are measured in Unicode characters rather than in bytes.

Known issues:
- some platforms are able to infer whether text is right-to-left from the text itself;
with other platforms, it's necessary to call `Fl_Native_Input::right_to_left(bool)`.
- member function `int Fl_Native_Input::copy_cuts()` is implemented only for Linux/Unix.
- no experiment done with Hebrew yet. Hopefully it should work OK. Feedback welcome.
- undo is limited to one level under Windows. That's what the native widget offers.

**Comment by ManoloFLTK** [comment]
@MoAlyousef Would you be willing to try the proposed new FLTK widget -`Fl_Native_Input`- and experiment with it in the perspective of using a right-to-left script in an FLTK app? That would be great.

**Comment by MoAlyousef** [comment]
Hi @ManoloFLTK 
Sure. I’ll give it a spin later this evening and report back.

**Comment by MoAlyousef** [comment]
Thank you for your effort @ManoloFLTK

I'll test things a bit more over the weekend. From a cursory trial this evening, the text widget works as expectedly on windows (tested on windows 11). 
On linux (ubuntu 22.04 on wsl, both X11 and wayland), I had issues in both rtl and ltr (English):
- For some reason, the cursor (insertion position) changes after the first letter for some reason. This occurs in English as well.
- Shortcuts don't work as well. Using Shift and the arrow keys doesn't select, neither do the Home or End keys (or Shift + Home & Shift + End).

I've done a screen recording on the behavior on both systems, windows:

https://github.com/user-attachments/assets/11eb3e67-2228-41f3-ae0a-c5719827ecc5

WSL:

https://github.com/user-attachments/assets/36f674dd-ab93-4f13-834b-17f2f4317fc7

I can test over the weekend on a macos 12 and on a proper linux system (kubuntu).

**Reply by ManoloFLTK** [reply]
Thanks. I see all the issues you report and will try to fix them. None of these issues occur under macOS.

**Reply by ManoloFLTK** [reply]
Reported issues should all be fixed after 
```
git pull
make native-text-widget
```

**Comment by MatthiasWM** [comment]
Thanks for this very nice addition. I am OK with adding this to FLTK when it works well, but - without having seen the code - there are some general things I learned from initially writing Sys Menu Bar, which @ManoloFLTK kept improving so well over the the years.

There is also the wxWidget library that exclusively wraps native widgets in a cross platform way. App devs like it because it looks native to the end user, but it's also limiting because it always only implements the lowest common denominator. I understand that editing and rendering text correctly is pretty much impossible to implement, so I see that a native text editor makes sense, but we still must offer as-good-as-possible editing without the native alternative.

1. adding features always requires additional maintenance, especially if a native interface is so new to FLTK and can't be easily maintained by anyone but the author
2. the widget must always work, even if the native equivalent is not available at runtime. We can't expect app devs to write complex #ifdef statements, just in case a user wants to compile their app on some obscure distro
3. the widget should look like FLTK as much as possible. Its colors and fonts should be the same as FLTK, and it would be good to have the same or similar boxtype. I know that we will not get the scroll bars to match, but that's what we have to live with
4. the API must be FLTK compatible, which is another way of item 2. It must be easy for the user, to switch between a native widget and an FLTK widget, preferable just with a single #ifdef
5. the native widget must resize, scale, and scroll, just like an FLTK widget would
6. there should be no (as little as possible) code in unrelated FLTK areas. A native widget should still work on its own without bodging the FLTK core and generating code paths we will not understand later
7. there will be FLTK versions for Android, iOS, and WebASM, maybe embedded. This is yet another way to emphasize item 2: we must always be able to fall back to FLTK if the native resource is not available

I am sure I could fill to 10 items, but I got to go now. Thanks, @ManoloFLTK , for implementing this. I am looking forward to testing.

**Reply by ManoloFLTK** [reply]
>  adding features always requires additional maintenance, especially if a native interface is so new to FLTK and can't be easily maintained by anyone but the author

We're also in this situation with the Wayland backend. I wrote an extensive description of the code inner workings to help others dive in it.

>    the widget must always work, even if the native equivalent is not available at runtime. We can't expect app devs to write complex #ifdef statements, just in case a user wants to compile their app on some obscure distro

Now the code compiles on any platform, but produces empty frames when the native widget is not available. I will change that so native-less platforms produce `Fl_Input` and `Fl_Multi_Line_Input` widgets instead.

>    the widget should look like FLTK as much as possible. Its colors and fonts should be the same as FLTK, and it would be good to have the same or similar boxtype. I know that we will not get the scroll bars to match, but that's what we have to live with

>    the API must be FLTK compatible, which is another way of item 2. It must be easy for the user, to switch between a native widget and an FLTK widget, preferable just with a single #ifdef

 >   the native widget must resize, scale, and scroll, just like an FLTK widget would

all of these are already in place

>    there should be no (as little as possible) code in unrelated FLTK areas. A native widget should still work on its own without bodging the FLTK core and generating code paths we will not understand later

No change to the FLTK core code is impossible. The long text above details what it does and where it is.

>    there will be FLTK versions for Android, iOS, and WebASM, maybe embedded. This is yet another way to emphasize item 2: we must always be able to fall back to FLTK if the native resource is not available

These would initially use `Fl_Input` until their own native widget would be added.

I look forward to your tests of the new proposed widget.

**Reply by MoAlyousef** [reply]
Thanks @ManoloFLTK 
I've pulled the changes. On linux, the cursor pos issue seems fixed.
I have an issue with arrow movements in latin characters, the left and right arrow keys seem inverted. In Arabic this works as expected.
Regarding Home and End, it seems that since I have them on my numpad, it's not recognized as shortcuts.

**Reply by ManoloFLTK** [reply]
> I have an issue with arrow movements in latin characters, the left and right arrow keys seem inverted. In Arabic this works as expected.

@MoAlyousef That is a question where your expertise is especially useful for me. What is implemented under Linux is the same as what I see under Windows: the left arrow moves left which means advances in Arabic text, and keeps advancing in Latin text interspersed in Arabic text which means moves right. At the end of the Latin text, back in Arabic text, the left arrow moves the cursor to the left again. Under Windows, that is entirely done by the OS. FLTK doesn't pilot the cursor. Such movements make sense to me. Under macOS, the left arrow moves left always, and there also FLTK doesn't intervene, only macOS is at work. Overall, isn't having Linux behave as Windows correct?

**Reply by ManoloFLTK** [reply]
> Regarding Home and End

and also regarding arrow keys, things should be better after another `git pull`.

**Reply by MoAlyousef** [reply]
Scratch the point about arrow movement. I had forgotten to toggle RtoL to LtoR.
Regarding the Home and End, I pulled again but I'm still not sure what the issue might be. I'll try to dig deeper on my side as well.
The Shift + Arrow keys now works to select text, however the behavior is a bit strange in both Latin and Arabic:

https://github.com/user-attachments/assets/7534a4f4-6640-46cf-9e72-927524be74e4

For example if you move to mid word and Shift + Right it will select the letter on the right, if you then press Shift + Left, it will not return the cursor, it will expand the selection on the left.
On windows, this works correctly.

**Reply by ManoloFLTK** [reply]
> The Shift + Arrow keys now works to select text, however the behavior is a bit strange in both Latin and Arabic

That should be fixed after yet another `git pull`.

**Comment by ManoloFLTK** [comment]
@MatthiasWM 
> the widget must always work, even if the native equivalent is not available at runtime. We can't expect app devs to write complex #ifdef statements, just in case a user wants to compile their app on some obscure distro

Commit ec172d2 of the `text_widget` branch adds class `Fl_Backup_Native_Input_Driver` which provides a backup widget when a platform doesn't implement (yet?) a true native text widget. The net effect of this class is that `Fl_Native_Input` is implemented as an `Fl_Group` containing an `Fl_Input` (or `Fl_Multiline_Input`) and that focus is handled as FLTK handles focus normally.

Thus, no `#ifdef`at all is needed. The FLTK library automatically uses a text widget available on the running platform when an app uses class `Fl_Native_Input`.

FLTK 1.4 draws with Cairo by default under X11 and therefore uses the new native text widget which requires GTK3 and Cairo. But some old X11 platforms may lack Cairo or GTK3. These would use the backup text widget.

**Comment by Albrecht-S** [comment]
@ManoloFLTK Thanks for your work on this new widget for RTL input. I didn't have a chance to test it yet but I suggest to change the names by replacing "Text_Widget" (which sounds inappropriate) with "Input" in analogy to `Fl_Input` and `Fl_Multiline_Input`. Thus I propose the following names:
- `Fl_Native_Text_Widget` -> `Fl_Native_Input` ( <-> `Fl_Input`)
- `Fl_Native_Multiline_Text_Widget` -> `Fl_Native_Multiline_Input` ( <-> `Fl_Multiline_Input`)

**Reply by ManoloFLTK** [reply]
@Albrecht-S Good suggestion. Done at d216cb1.

**Comment by Albrecht-S** [comment]
> no configure/make support yet.

I wouldn't bother to add it. Either `Fl_Native_Input` will be added in 1.5 or - if we really added it in 1.4.x - CMake support should be sufficient. Our announcement at the end of `configure` states: _"Some new FLTK build options will only be supported by CMake based builds."_ This could be one such new build option.

**Comment by MoAlyousef** [comment]
I've pulled the latest source and also tried on macos, windows, and linux. Honestly I can't find issues with it. Even in mixed text in multiline input.
The only issue I've run into was also in Linux with the X11 backend, when resizing the window to a smaller size, I get:
```
(native-input:1552667): Gtk-CRITICAL **: 09:29:27.558: gtk_box_gadget_distribute: assertion 'size >= 0' failed in GtkScrollbar
X_CreatePixmap: BadValue (integer parameter out of range for operation) 0x0
X_CopyArea: BadDrawable (invalid Pixmap or Window parameter) 0x6004a5
X_FreePixmap: BadPixmap (invalid Pixmap parameter) 0x6004a5
X_CreatePixmap: BadValue (integer parameter out of range for operation) 0x0
X_CopyArea: BadDrawable (invalid Pixmap or Window parameter) 0x6004aa
X_FreePixmap: BadPixmap (invalid Pixmap parameter) 0x6004aa

(native-input:1552667): Gtk-WARNING **: 09:29:27.577: gtk_widget_size_allocate(): attempt to allocate widget with width -3 and height 96
```

**Reply by ManoloFLTK** [reply]
@MoAlyousef Many thanks for all your testing. Validation of the new widget by a user of the Arabic language is very important. I will try to fix also the issue with resize.

**Reply by ggarra13** [reply]
@ManoloFLTK Great work, indeed!

**Comment by ggarra13** [comment]
@ManoloFLTK I tried it briefly on Linux and I liked it.  For @MoAlyousef, how do you handle mixed input in the new widgets?  For example, you type python code with left to right and then you want to switch for what is inside the print statement to Arabic: 

`print("arabic text")`

Is the switching possible and consistent on every platform?

**Reply by MoAlyousef** [reply]
@ggarra13  I've tried mixed input on windows 11, macos 12 and ubuntu 22.04 (on wsl, with both X11 and wayland backends), things were consistent on all tested platforms.
It just involves switching input methods (keyboard), which I do via an Alt-Shift shortcut for example on windows:

https://github.com/user-attachments/assets/5553fa98-7982-4b89-bfc4-2900bd5b852f

**Comment by ManoloFLTK** [comment]
@MoAlyousef 
> The only issue I've run into was also in Linux with the X11 backend, when resizing the window to a smaller size…

I don't succeed here in getting errors when I downsize the window. Could you identify a series of operations that triggers an error in your hands and post it here?

**Reply by MoAlyousef** [reply]
I've tried reproducing the error but can't 😅
I guess restarting my pc fixed something with wsl's xserver. I tried changing `window->resizable(window)` in the native-input.cxx, and when resizing to a small size I get:
```
(native-input:469217): Gtk-CRITICAL **: 22:58:23.892: gtk_box_gadget_distribute: assertion 'size >= 0' failed in GtkScrollbar
```

**Reply by ManoloFLTK** [reply]
@MoAlyousef Under Ubuntu, I see a different error when downsizing the window to very small sizes. A new commit removes that error. It might also fix the error you saw under wsl.
